rowhammer_tester/scripts/rowhammer.py

```diff
@@ -27,6 +27,69 @@ def get_expected_execution_cycles(payload):
             cycles += instr.timeslice - 1
     return cycles
 
+# returns the number of refreshes issued
+def encode_one_loop(*, unrolled, rolled, row_sequence, timings, encoder,
+        bank, refresh_op, payload, refresh=False, verbose=True):
+    tras = timings.tRAS
+    trp = timings.tRP
+    trefi = timings.tREFI
+    trfc = timings.tRFC
+    local_refreshes = 1
+    payload.append(encoder.I(refresh_op, timeslice=trfc))
+    # Accumulate an extra cycle for the jump at the end to be conservative
+    accum = trfc + 1
+    for idx in range(unrolled):
+        for row in row_sequence:
+            if accum + tras + trp > trefi:
+                payload.append(encoder.I(refresh_op, timeslice=trfc))
+                # Invariant: time between the beginning of two refreshes
+                # is is less than tREFI.
+                accum = trfc
+                local_refreshes += 1
+            accum += tras + trp
+            payload.extend([
+                encoder.I(OpCode.ACT,  timeslice=tras, address=encoder.address(bank=bank, row=row)),
+                encoder.I(OpCode.PRE,  timeslice=trp, address=encoder.address(col=1 << 10)),  # all
+            ])
+    jump_target = 2 * unrolled * len(row_sequence) + local_refreshes
+    assert jump_target < 2**Decoder.LOOP_JUMP
+    payload.append(encoder.I(OpCode.LOOP, count=rolled, jump=jump_target))
```

Comment:
You probably have off-by-one error. The instructions inside a loop body are executed `count+1` times as noted [here](https://github.com/antmicro/litex-rowhammer-tester/blob/3a72ceb7197604f01f2604ce6415d00e59253aee/rowhammer_tester/gateware/payload_executor.py#L50-L51) (so `count=0` is a pass-through with single execution of loop body). This way `local_refreshes * rolled` below is different than actual number of refreshes issued.